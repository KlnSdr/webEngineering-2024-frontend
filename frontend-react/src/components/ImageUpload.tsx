import React, { ChangeEvent, useEffect, useState } from "react";
import Button from "react-bootstrap/Button";

interface ImageUploadProps {
  initialValue: string | null;
  onChange: (img: string | null) => void;
}

const ImageUpload: React.FC<ImageUploadProps> = ({initialValue, onChange,}) => {
  const inputRef = React.useRef<HTMLInputElement>(null);
  const [selectedImage, setSelectedImage] = useState<string | null>(null);
  useEffect(() => {
    setSelectedImage(initialValue);
    if (inputRef.current) {
      inputRef.current.value = "";
    }
  }, [initialValue]);

  const handleImageChange = (event: ChangeEvent<HTMLInputElement>) => {
    const file = event.target.files?.[0];
    if (file) {
      const reader = new FileReader();
      reader.onloadend = () => {
        setSelectedImage(reader.result as string);
        onChange(reader.result as string);
      };
      reader.readAsDataURL(file);
    }
  };

  const handleRemoveImage = () => {
    setSelectedImage(null);
    if (inputRef.current) {
      inputRef.current.value = "";
    }
    onChange(null);
  };

  return (
    <div>
      <label htmlFor={"image-upload"}>WÃ¤hle ein Bild aus:</label>
      <input
        type="file"
        accept="image/*"
        onChange={handleImageChange}
        ref={inputRef}
        id={"image-upload"}
      />
      {selectedImage && (
        <div>
          <h2>Image Preview:</h2>
          <img
            src={selectedImage}
            alt="Selected"
            style={{ width: "300px", height: "auto" }}
          />
          <Button onClick={handleRemoveImage}>Remove Image</Button>
        </div>
      )}
    </div>
  );
};

export default ImageUpload;
