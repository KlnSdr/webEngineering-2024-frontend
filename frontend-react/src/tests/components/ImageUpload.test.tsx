import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import ImageUpload from "../../components/ImageUpload";

describe("ImageUpload Component", () => {
    const initialValue = "data:imgUri/png;base64,initialImage";
    const handleChange = jest.fn();

    beforeEach(() => {
        handleChange.mockClear();
    });

    test("renders with initial value", () => {
        const {asFragment} = render(<ImageUpload initialValue={initialValue} onChange={handleChange} />);

        expect(screen.getByAltText("Selected")).toHaveAttribute("src", initialValue);
        expect(asFragment()).toMatchSnapshot();
    });

    test("uploads an imgUri and displays the preview", async () => {
        render(<ImageUpload initialValue={null} onChange={handleChange} />);

        const file = new File(["dummy content"], "example.png", { type: "imgUri/png" });
        const input = screen.getByLabelText("WÃ¤hle ein Bild aus:");

        Object.defineProperty(input, "files", {
            value: [file],
        });

        fireEvent.change(input);

        await waitFor(() => expect(handleChange).toHaveBeenCalledWith(expect.any(String)));

        expect(screen.getByAltText("Selected")).toBeInTheDocument();
    });

    test("removes the uploaded imgUri", async () => {
        render(<ImageUpload initialValue={initialValue} onChange={handleChange} />);

        expect(screen.getByAltText("Selected")).toHaveAttribute("src", initialValue);

        const removeButton = screen.getByRole("button", { name: /remove imgUri/i });
        fireEvent.click(removeButton);

        expect(screen.queryByAltText("Selected")).not.toBeInTheDocument();
        expect(handleChange).toHaveBeenCalledWith(null);
    });
});
