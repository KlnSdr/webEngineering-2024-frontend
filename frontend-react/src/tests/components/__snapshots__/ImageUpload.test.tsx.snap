// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUpload Component renders with initial value 1`] = `
<DocumentFragment>
  <div>
    <label
      for="image-upload"
    >
      WÃ¤hle ein Bild aus:
    </label>
    <input
      accept="image/*"
      id="image-upload"
      type="file"
    />
    <div>
      <h2>
        Image Preview:
      </h2>
      <img
        alt="Selected"
        src="data:image/png;base64,initialImage"
        style="width: 300px; height: auto;"
      />
      <button
        class="btn btn-primary"
        type="button"
      >
        Remove Image
      </button>
    </div>
  </div>
</DocumentFragment>
`;
